hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventCounterPublisherContainer.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.wan.impl;
+
+import com.hazelcast.cache.impl.CacheService;
+import com.hazelcast.map.impl.MapService;
+
+/**
+ * Thread safe container for {@link WanEventCounterContainer}s grouped by
+ * WAN replication name and target group name.
+ */
+class WanEventCounterPublisherContainer {
```

Comment:
We should come up with a better name. This is a royal PITA to read.

Reply:
The problem starts with WanEventCounter.

This is not a single counter, it is a set of counters for multiple maps or caches.

Reply:
Can't we get rid of the WanEventCounterPublisherContainer and let the WanCounter (should be renamed) take care of finding the right counter based on service name and datastructure name?

Reply:
And we we need to get the counters for every call? Can't we just get those counters once and reuse the reference?

Reply:
The whole thing with the WanEventCounter is smelly.

Reply:
I'll add the caching of the counters. In what way do you consider the `WanEventCounter` smelly? The way I see it, the biggest issue you have with it is that they are part of some "container" objects. This comes from the requirement that a certain event counter should be:
- grouped by publisher. You want to know how many events were sent to one cluster but not the other
- grouped by map/cache. You want to know how many events were sent to one cluster for one map and how many were dropped for a specific cache.
- counters for a specific map/cache need to be removed once a map/cache is destroyed

The final outcome of these requirements is a diagnostics output like this:
```
28-05-2018 13:46:33 WAN[
                          syncState[
                                  status=READY
                                  activeWanConfigName=null
                                  activePublisherName=null
                                  syncedPartitionCount=0]
                          atob[
                                  B[
                                          outboundQueueSize=0
                                          totalPublishedEventCount=1,000
                                          totalPublishLatency=1,293,592
                                          connected=true
                                          paused=false
                                          mapSentEvents-myMap[
                                                  updateCount=1,000
                                                  removeCount=0
                                                  syncCount=0
                                                  droppedCount=0]
                                          mapSentEvents-myMap2[
                                                  updateCount=1,000
                                                  removeCount=0
                                                  syncCount=0
                                                  droppedCount=0]
                                          mapSentEvents-myMap3[
                                                  updateCount=1,000
                                                  removeCount=0
                                                  syncCount=0
                                                  droppedCount=0]]
                                  C[
                                          outboundQueueSize=0
                                          totalPublishedEventCount=1,000
                                          totalPublishLatency=1,293,674
                                          connected=true
                                          paused=false
                                          mapSentEvents-myMap[
                                                  updateCount=1,000
                                                  removeCount=0
                                                  syncCount=0
                                                  droppedCount=0]]]]

28-05-2018 13:46:33 WAN[
                          mapReceivedEvents-myMap[
                                  syncCount=0
                                  updateCount=1,000
                                  removeCount=0]
                          mapReceivedEvents-myMap2[
                                  syncCount=0
                                  updateCount=1,000
                                  removeCount=0]
                          mapReceivedEvents-myMap3[
                                  syncCount=0
                                  updateCount=1,000
                                  removeCount=0]
                          syncState[
                                  status=READY
                                  activeWanConfigName=null
                                  activePublisherName=null
                                  syncedPartitionCount=0]]
```

Can we achieve the same with probes? Do you see some better way of organising the `AtomicLong` counters in the code (not in container classes)?